// piral-cli/lib/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.forceOverwriteKeys = exports.frameworkKeys = exports.availableReleaseProviders = exports.availableBundlers = exports.bundlerKeys = exports.sourceLanguageKeys = exports.clientTypeKeys = exports.piletBuildTypeKeys = exports.piralPublishTypeKeys = exports.piralBuildTypeKeys = exports.fromKeys = exports.publishModeKeys = exports.schemaKeys = void 0;
exports.valueOfForceOverwrite = valueOfForceOverwrite;
exports.keyOfForceOverwrite = keyOfForceOverwrite;
const enums_1 = require("./common/enums");
const constants_1 = require("./common/constants");
exports.schemaKeys = ['v0', 'v1', 'v2', 'v3', 'mf', 'none'];
exports.publishModeKeys = ['none', 'basic', 'bearer', 'digest'];
exports.fromKeys = ['local', 'remote', 'npm'];
exports.piralBuildTypeKeys = [
    'all',
    'release',
    'emulator',
    'emulator-package',
    'emulator-sources',
    'emulator-website',
];
exports.piralPublishTypeKeys = ['release', 'emulator'];
exports.piletBuildTypeKeys = ['default', 'standalone', 'manifest'];
exports.clientTypeKeys = ['npm', 'pnpm', 'pnp', 'yarn', 'lerna', 'rush', 'bun'];
exports.sourceLanguageKeys = ['ts', 'js'];
exports.bundlerKeys = ['none', ...constants_1.bundlerNames];
exports.availableBundlers = [];
exports.availableReleaseProviders = [];
exports.frameworkKeys = [...constants_1.frameworkLibs];
exports.forceOverwriteKeys = Object.keys(enums_1.ForceOverwrite).filter((m) => typeof enums_1.ForceOverwrite[m] === 'number');
function valueOfForceOverwrite(key) {
    for (const forceOverwriteKey of exports.forceOverwriteKeys) {
        if (forceOverwriteKey === key) {
            return enums_1.ForceOverwrite[forceOverwriteKey];
        }
    }
    return enums_1.ForceOverwrite.no;
}
function keyOfForceOverwrite(value) {
    for (const forceOverwriteKey of exports.forceOverwriteKeys) {
        if (enums_1.ForceOverwrite[forceOverwriteKey] === value) {
            return forceOverwriteKey;
        }
    }
    return exports.forceOverwriteKeys[0];
}
