// kras/dist/server/core/recorder.js
import { randomUUID } from 'crypto';
import { EventEmitter } from 'events';
export class Recorder extends EventEmitter {
    maximum;
    enabled = true;
    requests = [];
    errors = [];
    messages = [];
    constructor(maximum) {
        super();
        this.maximum = maximum;
    }
    disable() {
        this.enabled = false;
    }
    hit(start, end, request, response) {
        if (this.enabled) {
            const requests = this.requests;
            const id = randomUUID();
            const item = {
                id,
                start,
                end,
                request,
                response,
            };
            if (requests.length === this.maximum) {
                requests.shift();
            }
            this.requests.push(item);
            this.emit('recorded-request', item);
        }
    }
    message(time, data) {
        if (this.enabled) {
            const id = randomUUID();
            const item = {
                id,
                time,
                ...data,
            };
            this.messages.push(item);
            this.emit('recorded-message', item);
        }
    }
    miss(start, end, request) {
        if (this.enabled) {
            const id = randomUUID();
            const item = {
                id,
                start,
                end,
                request,
            };
            this.errors.push(item);
            this.emit('recorded-miss', item);
        }
    }
}
