// kras/dist/server/injectors/store-injector.js
import { resolve } from 'path';
import { open } from '../helpers';
export default class StoreInjector {
    db;
    config;
    constructor(options, config) {
        const directory = options.directory || config.directory;
        const today = new Date().toJSON().substr(0, 10);
        this.config = options;
        this.db = open(resolve(directory, options.filename || `unknown-requests-${today}.json`));
    }
    getOptions() {
        return {
            file: {
                title: 'File location',
                description: 'Location of the file to store the unknown / unhandled requests.',
                type: 'text',
                value: this.db.file,
            },
        };
    }
    setOptions(options) {
        const { file } = options;
        if (file && file !== this.db.file) {
            this.db.switchTo(file);
        }
    }
    get name() {
        return 'store-injector';
    }
    get active() {
        return this.config.active;
    }
    set active(value) {
        this.config.active = value;
    }
    handle(request) {
        this.db.insert({
            request,
            response: {
                status: 404,
            },
        });
    }
}
