// piral-cli/lib/build/run-debug-pilet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const envs_1 = require("../common/envs");
let handler;
let bundler;
function run(root, targetDir, outDir, outFile, piralInstances, externals, importmap, entryModule, version, logLevel, args) {
    (0, envs_1.setStandardEnvs)({
        production: false,
        piralInstances,
        root,
    });
    return handler.create({
        root,
        piralInstances,
        entryModule,
        targetDir,
        outDir,
        outFile,
        externals,
        importmap,
        version,
        develop: true,
        sourceMaps: true,
        contentHash: true,
        minify: false,
        logLevel,
        watch: true,
        args,
    });
}
process.on('message', async (msg) => {
    const root = process.cwd();
    try {
        switch (msg.type) {
            case 'init':
                handler = require(msg.path);
                break;
            case 'bundle':
                if (bundler) {
                    await bundler.bundle();
                }
                break;
            case 'start':
                bundler = await run(root, msg.targetDir, msg.outDir, msg.outFile, msg.piralInstances, msg.externals, msg.importmap, msg.entryModule, msg.version, msg.logLevel, msg);
                if (bundler) {
                    bundler.onStart(() => {
                        process.send({
                            type: 'pending',
                        });
                    });
                    bundler.onEnd((result) => {
                        if (msg.hmr) {
                            process.send({
                                type: 'update',
                                outHash: result.hash,
                                outName: result.name,
                                args: {
                                    requireRef: result.requireRef,
                                    version: msg.version,
                                    root,
                                },
                            });
                        }
                    });
                    process.send({
                        type: 'done',
                        outDir: msg.outDir,
                    });
                }
                break;
        }
    }
    catch (error) {
        process.send({
            type: 'fail',
            error: error?.message,
        });
    }
});
