// piral-cli/lib/common/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.legacyCoreExternals = exports.declarationEntryExtensions = exports.bundlerNames = exports.entryModuleExtensions = exports.piralJsonSchemaUrl = exports.piletJsonSchemaUrl = exports.frameworkLibs = exports.emulatorWebsiteName = exports.emulatorSourcesName = exports.emulatorPackageName = exports.emulatorName = exports.emulatorJson = exports.releaseName = exports.allName = exports.defaultSchemaVersion = exports.piralBaseRoot = exports.filesOnceTar = exports.piletJson = exports.piralJson = exports.packageJson = exports.appName = exports.filesTar = exports.defaultRegistry = void 0;
exports.defaultRegistry = 'https://registry.npmjs.org/';
exports.filesTar = 'files';
exports.appName = 'piral';
exports.packageJson = 'package.json';
exports.piralJson = 'piral.json';
exports.piletJson = 'pilet.json';
exports.filesOnceTar = 'files_once';
exports.piralBaseRoot = 'piral-base/package.json';
exports.defaultSchemaVersion = 'v2';
exports.allName = 'all';
exports.releaseName = 'release';
exports.emulatorJson = 'emulator.json';
exports.emulatorName = 'emulator';
exports.emulatorPackageName = 'package';
exports.emulatorSourcesName = 'sources';
exports.emulatorWebsiteName = 'website';
exports.frameworkLibs = ['piral', 'piral-core', 'piral-base'];
exports.piletJsonSchemaUrl = 'https://docs.piral.io/schemas/pilet-v0.json';
exports.piralJsonSchemaUrl = 'https://docs.piral.io/schemas/piral-v0.json';
exports.entryModuleExtensions = ['.ts', '.tsx', '.js', '.jsx'];
exports.bundlerNames = [
    'bun',
    'esbuild',
    'parcel',
    'parcel2',
    'rollup',
    'rspack',
    'webpack',
    'webpack5',
    'vite',
    'vite5',
    'vite6',
    'xbuild',
    'netpack',
];
exports.declarationEntryExtensions = ['.html', '.pug', ...exports.entryModuleExtensions];
exports.legacyCoreExternals = [
    '@dbeining/react-atom',
    '@libre/atom',
    'history',
    'react',
    'react-dom',
    'react-router',
    'react-router-dom',
    'tslib',
    'path-to-regexp',
];
