// piral-cli-webpack/lib/actions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildPilet = exports.debugPilet = exports.buildPiral = exports.debugPiral = exports.watchPiral = void 0;
const path_1 = require("path");
const constants_1 = require("./constants");
exports.watchPiral = {
    path: (0, path_1.resolve)(__dirname, 'webpack', 'piral.js'),
};
exports.debugPiral = {
    flags(argv) {
        return argv
            .string('config')
            .describe('config', 'Sets configuration file for modifying the Webpack configuration.')
            .default('config', constants_1.defaultWebpackConfig)
            .number('hmr-port')
            .describe('hmr-port', 'Sets the port to be used for HMR for reloading the application.')
            .default('hmr-port', 62123);
    },
    path: (0, path_1.resolve)(__dirname, 'webpack', 'piral.js'),
};
exports.buildPiral = {
    flags(argv) {
        return argv
            .string('config')
            .describe('config', 'Sets configuration file for modifying the Webpack configuration.')
            .default('config', constants_1.defaultWebpackConfig);
    },
    path: (0, path_1.resolve)(__dirname, 'webpack', 'piral.js'),
};
exports.debugPilet = {
    flags(argv) {
        return argv
            .string('config')
            .describe('config', 'Sets configuration file for modifying the Webpack configuration.')
            .default('config', constants_1.defaultWebpackConfig);
    },
    path: (0, path_1.resolve)(__dirname, 'webpack', 'pilet.js'),
};
exports.buildPilet = {
    flags(argv) {
        return argv
            .string('config')
            .describe('config', 'Sets configuration file for modifying the Webpack configuration.')
            .default('config', constants_1.defaultWebpackConfig);
    },
    path: (0, path_1.resolve)(__dirname, 'webpack', 'pilet.js'),
};
