// piral-cli/lib/apps/declaration-piral.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.declarationPiralDefaults = void 0;
exports.declarationPiral = declarationPiral;
const path_1 = require("path");
const common_1 = require("../common");
const types_1 = require("../types");
exports.declarationPiralDefaults = {
    entry: './',
    target: './dist',
    forceOverwrite: common_1.ForceOverwrite.yes,
    logLevel: types_1.LogLevels.info,
};
async function declarationPiral(baseDir = process.cwd(), options = {}) {
    const { entry = exports.declarationPiralDefaults.entry, target = exports.declarationPiralDefaults.target, forceOverwrite = exports.declarationPiralDefaults.forceOverwrite, logLevel = exports.declarationPiralDefaults.logLevel, } = options;
    (0, common_1.ensure)('baseDir', baseDir, 'string');
    (0, common_1.ensure)('entry', entry, 'string');
    (0, common_1.ensure)('target', target, 'string');
    const fullBase = (0, path_1.resolve)(process.cwd(), baseDir);
    (0, common_1.setLogLevel)(logLevel);
    if (await (0, common_1.createPiralDeclaration)(fullBase, entry, target, forceOverwrite, logLevel)) {
        (0, common_1.logDone)(`Declaration created successfully in "${target}"!`);
    }
}
